USB: cdc_acm: use ringbuf library and fixes

finikorg opened the pull request:
Remove old artefacts and use ringbuf for RX path.

Fixes #14697
Fixes #14698
Might Fixes #14288

finikorg committed:
usb: cdc_acm: Make parameters boolean

Remove 'U' after every assignment and make code more logical.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Use ringbuf for RX path

Use ringbuf library for handling RX CDC ACM path. Remove old code
artifacts handling specific hardware.

Fixes #14288

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: cdc_acm_composite: Fix missing irq handling

If there are several queued interrupts we can miss some of them. Use
while() loop to catch them all.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-phytec-iot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -328,36 +327,18 @@ static void cdc_acm_bulk_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+	usb_read(ep, dev_data->rx_buf, bytes_to_read, &read);
+
+	if (!ring_buf_put(dev_data->rx_ringbuf, dev_data->rx_buf, read)) {
+		LOG_WRN("Rinf buffer full");
```
New data will be dropped? Also we can not block here as it could be called from ISR (drivers without worker or own thread). I would change it to LOG_ERR.

finikorg replied:
@jfischer-phytec-iot why do you think it can block? The new data would be dropped the same way it was done before, functionality should not change.

finikorg replied:
Changed to LOG_ERR()

jtguggedal commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -1075,9 +1042,11 @@ static const struct uart_driver_api cdc_acm_driver_api = {
 #endif /* CONFIG_USB_COMPOSITE_DEVICE */
 
 #define DEFINE_CDC_ACM_DEV_DATA(x)					\
+	RING_BUF_DECLARE(rx_ringbuf_##x, 512);				\
```
Is it an option to make the ring buffer size configurable in Kconfig? I tested the patch and ran into issues when the buffer was filled and started dropping data before the callback was served from the system workqueue.

finikorg replied:
I can make it configurable.

jtguggedal replied:
Thanks!

finikorg replied:
Done

finikorg commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -185,10 +185,10 @@ struct cdc_acm_dev_data_t {
-	u8_t rx_ready;                 /* Rx ready status */
-	u8_t tx_irq_ena;               /* Tx interrupt enable status */
-	u8_t rx_irq_ena;               /* Rx interrupt enable status */
+	s8_t tx_ready;
```
This is to eliminate 'U' after assignments.

jfischer-phytec-iot replied:
I suggest to use bool or bit field.

finikorg replied:
Changed to bool

finikorg committed:
usb: cdc_acm: Make ring buffer size configurable

Add new configuration parameter USB_CDC_ACM_RINGBUF_SIZE, default to
1024.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: console: Select console device name

Instead of relying on overlays which requires to write overlay for
every board (100+) define console device name.

Fixes #14698

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

masz-nordic commented:
Tested the throughput, `cdc_acm` without loopback and logs is at 312 kB/s.

jfischer-no reviewed (approved):
LGTM

jtguggedal commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -625,34 +607,20 @@ static int cdc_acm_fifo_fill(struct device *dev,
-		rx_data[i] = dev_data->rx_buf[(dev_data->rx_buf_tail + i) %
-					      CDC_ACM_BUFFER_SIZE];
-	}
+	len = MIN(size, ring_buf_space_get(dev_data->rx_ringbuf));
```
Should this rather be something like
`len = MIN(size, dev_data->rx_ringbuf->size);`
or 
`len = MIN(size, dev_data->rx_ringbuf->size - ring_buf_space_get(dev_data->rx_ringbuf));`

Currently, no data will be read out using `uart_fifo_read(...)`if the ring buffer is full, as `ring_buf_space_get(dev_data->rx_ringbuf)` returns 0 free space left in the ring buffer.

finikorg replied:
Thanks for noticing this, I will fix it.
probably we are not supposed to use `rx_ringbuf->size` directly, and size should be multiplied by 4 since `Size of buf in 32-bit chunks`. Maybe we can just use `size` directly.

jtguggedal replied:
That's true. I think you're right in that using `size` directly is fine.

jtguggedal commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -328,36 +328,18 @@ static void cdc_acm_bulk_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+	usb_read(ep, dev_data->rx_buf, bytes_to_read, &read);
+
+	if (!ring_buf_put(dev_data->rx_ringbuf, dev_data->rx_buf, read)) {
+		LOG_ERR("Rinf buffer full");
```
Tiny typo, should be `Ring buffer full`

finikorg replied:
fixed
